feat: add `math/base/special/sind`

anandkaranubc opened the pull request:
Resolves https://github.com/stdlib-js/stdlib/issues/32

## Description

> What is the purpose of this pull request?

This pull request:

- Adds the JS and C implementation of the function to compute the sine of `x`, where `x` is in degrees.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/stdlib/issues/32

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: add JS implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: add test files and their fixtures

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add C implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
bench: add benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
build: add dependencies

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: add documentation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc commented:
I just tweaked the implementation of `math/base/special/cosd` and modified it to align with the current requirements.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/test/test.js:
```diff
@@ -0,0 +1,123 @@
+	t.end();
+});
+
+tape( 'if provided a multiple of `180.0`, the function returns `0.0`', function test( t ) {
```
This is the test that has been added for this specific function.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+	double t;
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = ( 10.0 * rand_double() ) - 5.0;
+	}
```
The benchmarks have been updated to account for the ongoing refactoring.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+	var y;
+	var i;
+
+	x = uniform( 100, -5.0, 5.0 );
```
And here too :)

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/README.md:
```diff
@@ -0,0 +1,185 @@
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+</section>
```
I have removed all the links from here, following the same approach used for `hyp2f1`.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/math/base/special/sind.h"
+#include "stdlib/math/base/napi/unary.h"
+
+// cppcheck-suppress shadowFunction
```
Added this line to suppress the warning. This line isn't present in `cosd`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,70 @@
+# Extract the directory in which this file resides:
+dir = dirname( file );
+
+# Generate fixture data for negative values:
+x = range( -10.0, stop = 0, length = 1000 );
+gen( x, "negative.json" );
+
+# Generate fixture data for positive values:
+x = range( 10.0, stop = 0, length = 1000 );
+gen( x, "positive.json" );
```
I have used the same fixtures that were used for `cosd`

kgryte replied:
So long as you trust them... 😑

kgryte replied:
I never reviewed the `cosd` implementation, so I suggest treating the `cosd` package with some suspicion.

anandkaranubc replied:
Bad reference point ig.

anandkaranubc replied:
Seems like a major change to be made. Even the `secd` uses a similar implementation

kgryte replied:
💆

kgryte replied:
🧘

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+		return NaN;
+	}
+
+	if ( isInteger( x / 180.0 ) ) {
+		return 0.0;
+	}
```
Just curious, the only reason we add this is to save some computation time, right?

kgryte replied:
As referenced in the original issue, the implementation is actually more involved: https://github.com/JuliaLang/julia/blob/d63adeda50d5b8c440a7c29c9a65357a98fe927f/base/special/trig.jl#L1183.

kgryte replied:
This `cosd` PR should never have been merged: https://github.com/stdlib-js/stdlib/commit/0b0e240f6959ac1dae3bc00fe42b9930eb887e93.

anandkaranubc replied:
I see! I was just about to ask about this.

kgryte replied:
The purpose of these `*d` implementations is to provide higher accuracy results when working with degrees, rather than radians.

anandkaranubc replied:
And this change is consistent with all the other `*d` implementations that have been merged?

anandkaranubc committed:
refactor: update sind function to improve modular reduction and edge case handling

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,70 @@
+dir = dirname( file );
+
+# Generate fixture data for negative values:
+x = range( -360.0, stop = 0, length = 1000 );
+gen( x, "negative.json" );
+
+# Generate fixture data for positive values:
+x = range( 0.0, stop = 360.0, length = 1000 );
+gen( x, "positive.json" );
```
Thoroughly test the function over the range [-360, 360].

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/lib/main.js:
```diff
@@ -0,0 +1,101 @@
+		return NaN;
+	}
+
+	rx = fmod( x, 360.0 );
```
This can easily be replaced by `x % 360.0` in JS. The use of `fmod` ensures consistency between the C and JS implementations.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/src/main.c:
```diff
@@ -0,0 +1,75 @@
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/sind.h"
+#include "stdlib/math/base/special/kernel_sin.h"
+#include "stdlib/math/base/special/kernel_cos.h"
+#include "stdlib/math/base/special/deg2rad.h"
+#include "stdlib/math/base/special/signum.h"
+#include "stdlib/math/base/special/abs.h"
+#include "stdlib/math/base/special/fmod.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/assert/is_infinite.h"
```
Does the order of the `#include` statements need to be changed here?

kgryte replied:
No. We don't impose any order of `stdlib` includes atm.

anandkaranubc committed:
chore: fix the spacing

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/lib/main.js:
```diff
@@ -0,0 +1,101 @@
+		return signum( rx ) * kernelCos( xRad, 0.0 );
+	}
+	if ( arx === 180.0 ) {
+		return signum( rx ) * 0.0;
```
This line can be changed to `return 0.0`. Any thoughts on which one to keep?

kgryte replied:
Can it? I don't think it can. You need to correctly return signed zero.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/src/main.c:
```diff
@@ -0,0 +1,75 @@
+		return stdlib_base_signum( rx ) * stdlib_base_kernel_cos( xRad, 0.0 );
+	}
+	if ( arx == 180.0 ) {
+		return stdlib_base_signum( rx ) * 0.0;
```
Same here :)

kgryte replied:
Signed zero?

anandkaranubc replied:
You are right. Sorry, my bad.  

I was under the wrong impression that multiplying by `0.0` would turn signed zero into `0.0`.

anandkaranubc replied:
Time to study more about signed zero....

anandkaranubc commented:
Once this is approved, I will also update the other implementations that have been merged already.

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into feat/sind

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sind/src/main.c:
```diff
@@ -0,0 +1,75 @@
+		return stdlib_base_signum( rx ) * 0.0;
+	}
+	if ( arx < 225.0 ) {
+		xRad = stdlib_base_deg2rad( ( 180.0-arx ) * stdlib_base_signum( rx ) );
+		return stdlib_base_kernel_sin( xRad, 0.0 );
```
```suggestion
		return stdlib_base_kernel_sin( stdlib_base_deg2rad( ( 180.0-arx )*stdlib_base_signum( rx ) ), 0.0 );
```

@gunjjoshi, do you think the same kind of refactoring also applies in here?

anandkaranubc replied:
There are some parts of the C and JS implementations where `xRad` can be inlined while keeping the expressions simple. However, cases like the one above make me question whether the use of `xRad` is appropriate here and if it "could" be written in every `if` block to maintain a consistent structure.

gunjjoshi replied:
@anandkaranubc You're right, although this case isn't as simple as the previous one, but, I think, let's follow the same refactoring here too, so that it matches with `cosd`.

anandkaranubc replied:
Sure! On it rn.... Thanks!

anandkaranubc replied:
@gunjjoshi Done! To be honest, it doesn't look that "cluttered".

gunjjoshi replied:
Agreed, thanks!

anandkaranubc committed:
chore: remove xRad for cleaner code

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter reviewed (approved):
Lgtm.  Thanks, @anandkaranubc .
